File: array/golf_bot/golf_bot.py
import numpy as np
from numpy.fft import rfft, irfft


def self_convolution(arr):
    n = len(arr)
    k = 2 * n - 1
    ftarr = rfft(arr, k)
    return irfft(ftarr * ftarr, k)


def calc_holes(moves, distances):
    moves_poly = np.zeros(np.max(moves) + 1)
    for move in moves:
        moves_poly[move] = 1
    moves_poly[0] = 1
    conv = self_convolution(moves_poly)
    return sum(round(conv[d]) > 0 for d in distances)


moves = [1, 3, 5]
distances = [2, 4, 5, 7, 8, 9]

expected_out = 4
print(calc_holes(moves, distances))

File: array/golf_bot/golf_bot.jl

function calc_holes(moves::Vector{Int}, distances::Vector{Int})
    move_set = Set(moves)
    M = 2 * maximum(moves) - 1
    moves_poly = [ d in move_set for d in 0:M]
    moves_poly[1] = 1
    moves_transform = fft(moves_poly)
    conv = ifft(moves_transform .* moves_transform)
    return sum(round(real(conv[d + 1])) > 0 for d in distances)
end

moves = [1, 3, 5]
distances = [2, 4, 5, 7, 8, 9]
expected_out = 4
println(calc_holes(moves, distances))
